stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -760,25 +761,33 @@ extension ${Self} {
 extension ${Self} {
 % for srcBits in allFloatBits:
 %   That = floatName[srcBits]
-%   if Self != That:
 
-%     if srcBits == 80:
+%   if srcBits == 80:
 #if !os(Windows) && (arch(i386) || arch(x86_64))
-%     end
+%   end
 
   /// Construct an instance that approximates `other`.
   public init(_ other: ${That}) {
-%     if srcBits > bits:
+%   if srcBits > bits:
     _value = Builtin.fptrunc_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
-%     else:
+%   elif srcBits < bits:
     _value = Builtin.fpext_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
-%     end
+%   else:
+    _value = other._value
+%   end
+  }
+  
+  /// Create an ${That} initialized to `value`, if `value` can be represented without rounding.
+  @_transparent
+  public init?(exactly other: ${That}) {
+    self.init(other)
+    if ${That}(self) != other {
+      return nil
+    }
   }
```

Comment:
Could you simplify this implementation for the case when `srcBits <= bits`?  I don't think you need to check in that case.